docroot/sites/classrooms.uiowa.edu/modules/classrooms_core/src/BatchRooms.php

```diff
@@ -0,0 +1,321 @@
+<?php
+
+namespace Drupal\classrooms_core;
+
+use Drupal\Core\Entity\FieldableEntityInterface;
+
+/**
+ * Batch processes for importing rooms data.
+ */
+class BatchRooms {
+
+  /**
+   * Batch process callback.
+   *
+   * @param int $batch_id
+   *   Id of the batch.
+   * @param array $nodes
+   *   Individual nodes to be processed.
+   * @param object $context
+   *   Context for operations.
+   */
+  public static function processNode(int $batch_id, array $nodes, object &$context) {
+    // Optional message displayed under the progressbar.
+    $context['message'] = t('Running Batch "@id"', [
+      '@id' => $batch_id,
+    ]);
+
+    foreach ($nodes as $node) {
+
+      if (!$node instanceof FieldableEntityInterface) {
+        continue;
+      }
+
+      if ($node->hasField('field_room_building_id') &&
+        !$node->get('field_room_building_id')->isEmpty() &&
+        $node->hasField('field_room_room_id') &&
+        !$node->get('field_room_room_id')->isEmpty()
+      ) {
+        $building_id = $node->get('field_room_building_id')?->getString();
+        $room_id = $node->get('field_room_room_id')->value;
+      }
+      else {
+        continue;
+      }
+
+      $updated = FALSE;
+
+      // Grab MAUI room data.
+      $data = \Drupal::service('uiowa_maui.api')
+        ->getRoomData($building_id, $room_id);
+
+      // If we weren't able to get data for this room,
+      // log a notice and move on to the next one.
+      if (!$data) {
+        $context['message'] = t('No data found for @building @room', [
+          '@building' => $building_id,
+          '@room' => $room_id,
+        ]);
+        continue;
+      }
+
+      // If existing, update values if different.
+      // Comparing the Max Occupancy field
+      // to the maxOccupancy value from endpoint.
+      if ($node->hasField('field_room_max_occupancy') && isset($data[0]->maxOccupancy)) {
+        if (filter_var($data[0]->maxOccupancy, FILTER_VALIDATE_INT) !== FALSE) {
+          if ($node->get('field_room_max_occupancy')->value !== $data[0]->maxOccupancy) {
+            $updated = TRUE;
+            \Drupal::database()
+              ->update('node__field_room_max_occupancy')
+              ->fields([
+                'field_room_max_occupancy_value' => $data[0]->maxOccupancy,
+              ])
+              ->condition('revision_id', $node->getRevisionId(), '=')
+              ->execute();
+          }
+        }
+      }
```

Comment:
I'm trying to understand the rationale for doing this vs. just using `$node->set(...); $node->save();`. Is the intention here to piggyback on the previous node revision and not trigger another revision?

Reply:
The way things are currently set up, there's a presave that does a lot of this in a similar manner for when rooms are created (or if, say, they NEED to trigger a room to re-import for any reason). Setting and saving the node sets off the presave process again. The piggybacking off the previous revision is a side benefit, but not necessary.

Originally this was set up to call the API, check for differences, and if any were found, set a new revision and save the node. Then the presave handled the actual updates. The shift to db inserts came from a desire to not double-up on all the API calls and value checks in two places.

Reply:
https://github.com/uiowa/uiowa/pull/6370#issuecomment-1518344910

Reply:
the insert work isn't going as smoothly as I thought and we ultimately have similar node->save() code in the presave().

a node->save in a drush command passes through presave. i was trying to think of a way to abort full save if coming from drush and nothing has changed.

or maybe it should move out of the presave and instead be a local task to trigger an import for that node using a single version of the drush code.

Reply:
Could we so something like check the user of the revision in presave, and if it's uid 1, skip the extra api processing?

Reply:
As I said in EoD, I'd like to avoid using uid 1 as our check. We have logged in as that user to do things for clients and I don't want to add conditionals to that.

`$_SERVER["argv"]` exists when coming from a `$node->save()` in the batch process and could be used as an identifier within presave of what is saving it.

<img width="395" alt="Screenshot 2023-04-26 at 4 53 44 PM" src="https://user-images.githubusercontent.com/4663676/234712120-e8e1a869-eb5f-4b5e-b18a-5550b75cb1ea.png">

Reply:
Also noting the other idea of doing something like `$node->set('batch_update', TRUE)` in the batch process and then checking that value later in presave before doing additional processing.